File: android/src/main/kotlin/app/capgo/verisoul/VerisoulPlugin.kt
package app.capgo.verisoul

import ai.verisoul.sdk.VerisoulException
import ai.verisoul.sdk.helpers.webview.VerisoulSessionCallback
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import android.view.MotionEvent

@CapacitorPlugin(name = "Verisoul")
class VerisoulPlugin : Plugin() {

    private val implementation by lazy { VerisoulBridge(context) }

    @PluginMethod
    fun configure(call: PluginCall) {
        val environment = call.getString("environment")
        val projectId = call.getString("projectId")

        if (environment.isNullOrBlank()) {
            call.reject("environment is required", VerisoulErrorCodes.INVALID_ENVIRONMENT)
            return
        }
        if (projectId.isNullOrBlank()) {
            call.reject("projectId is required")
            return
        }

        try {
            implementation.configure(environment, projectId)
            call.resolve()
        } catch (exception: Throwable) {
            rejectVerisoul(call, "SDK configuration failed", exception)
        }
    }

    @PluginMethod
    fun getSessionId(call: PluginCall) {
        try {
            implementation.getSessionId(object : VerisoulSessionCallback {
                override fun onFailure(exception: Throwable) {
                    rejectVerisoul(call, "Failed to retrieve session ID", exception)
                }

                override fun onSuccess(sessionId: String) {
                    val ret = JSObject().apply {
                        put("sessionId", sessionId)
                    }
                    call.resolve(ret)
                }
            })
        } catch (exception: Throwable) {
            rejectVerisoul(call, "Failed to retrieve session ID", exception)
        }
    }

    @PluginMethod
    fun reinitialize(call: PluginCall) {
        try {
            implementation.reinitialize()
            call.resolve()
        } catch (exception: Throwable) {
            rejectVerisoul(call, "Failed to reinitialize SDK", exception)
        }
    }

    @PluginMethod
    fun recordTouchEvent(call: PluginCall) {
        val x = call.getFloat("x")
        val y = call.getFloat("y")
        val action = parseAction(call.getString("action"), call.getInt("action"))

        if (x == null || y == null || action == null) {
            call.reject("x, y, and action are required")
            return
        }

        try {
            implementation.recordTouchEvent(x, y, action)
            call.resolve()
        } catch (exception: Throwable) {
            rejectVerisoul(call, "Failed to record touch event", exception)
        }
    }

    private fun parseAction(actionString: String?, actionInt: Int?): Int? {
        return when (actionString) {
            "down" -> MotionEvent.ACTION_DOWN
            "up" -> MotionEvent.ACTION_UP
            "move" -> MotionEvent.ACTION_MOVE
            else -> when (actionInt) {
                0 -> MotionEvent.ACTION_DOWN
                1 -> MotionEvent.ACTION_UP
                2 -> MotionEvent.ACTION_MOVE
                else -> null
            }
        }
    }

    private fun rejectVerisoul(call: PluginCall, prefix: String, exception: Throwable) {
        val code = (exception as? VerisoulException)?.code ?: errorCodeFromMessage(exception)
        call.reject("$prefix: ${exception.message}", code, exception)
    }

    private fun errorCodeFromMessage(exception: Throwable): String {
        val message = exception.message.orEmpty().uppercase()
        return when {
            message.contains("ENVIRONMENT") -> VerisoulErrorCodes.INVALID_ENVIRONMENT
            message.contains("WEBVIEW") -> VerisoulErrorCodes.WEBVIEW_UNAVAILABLE
            message.contains("SESSION") -> VerisoulErrorCodes.SESSION_UNAVAILABLE
            else -> "UNKNOWN_ERROR"
        }
    }
}

File: ios/Sources/VerisoulPlugin/VerisoulPlugin.swift
import Foundation
import Capacitor

@objc(VerisoulPlugin)
public class VerisoulPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "VerisoulPlugin"
    public let jsName = "Verisoul"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "configure", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSessionId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reinitialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "recordTouchEvent", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = VerisoulBridge()

    @objc func configure(_ call: CAPPluginCall) {
        guard let environment = call.getString("environment") else {
            call.reject("environment is required", "INVALID_ENVIRONMENT")
            return
        }
        guard let projectId = call.getString("projectId"), !projectId.isEmpty else {
            call.reject("projectId is required")
            return
        }

        do {
            try implementation.configure(environment: environment, projectId: projectId)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, errorCode(for: error), error)
        }
    }

    @objc func getSessionId(_ call: CAPPluginCall) {
        Task {
            do {
                let sessionId = try await implementation.getSessionId()
                call.resolve(["sessionId": sessionId])
            } catch {
                call.reject(error.localizedDescription, errorCode(for: error), error)
            }
        }
    }

    @objc func reinitialize(_ call: CAPPluginCall) {
        implementation.reinitialize()
        call.resolve()
    }

    @objc func recordTouchEvent(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func errorCode(for error: Error) -> String {
        let description = error.localizedDescription.uppercased()
        if description.contains("ENVIRONMENT") {
            return "INVALID_ENVIRONMENT"
        }
        if description.contains("WEBVIEW") {
            return "WEBVIEW_UNAVAILABLE"
        }
        if description.contains("SESSION") {
            return "SESSION_UNAVAILABLE"
        }
        return "UNKNOWN_ERROR"
    }
}
